src/service/viewport-impl.js

```diff
@@ -1065,6 +1070,156 @@ export class ViewportBindingNaturalIosEmbed_ {
 }
 
 
+export class ViewportBindingNaturalIosScrollEmbed_
+    extends ViewportBindingNaturalIosEmbed_ {
+  /**
+   * @param {!Window} win
+   */
+  constructor(win) {
+    super(win);
+
+    // TODO
+    this.onScroll();
+
+    this.onResize(() => this.onResized_());
+  }
+
+  /** @override */
+  requiresFixedLayerTransfer() {
+    return false;
+  }
+
+  /** @override */
+  requiresFixedLayerLookup() {
+    return true;
+  }
+
+  /** @private */
+  setup_() {
+    const doc = this.win.document;
+    const documentElement = doc.documentElement;
+    const body = doc.body;
+
+    // Setup scrolling styles.
+    // html {
+    //   height: 100% !important;
+    //   -webkit-overflow-scrolling: touch;
+    // }
+    // body {
+    //   height: 100% !important;
+    //   overflow-x: hidden;
+    // }
+    setStyles(documentElement, {
+      height: '100% !important',
+      webkitOverflowScrolling: 'touch',
+    });
+    setStyles(body, {
+      height: '100% !important',
+      overflowX: 'hidden',
+    });
+
+    let hasContainer = body.children === 1;
+    if (hasContainer) {
+      for (let child = body.firstChild; child; child = child.nextSibling) {
+        // Does this Text Node contain any non-whitespace chars?
+        if (child.nodeType === /* Text */ 3 && /\S/.test(child.data)) {
+          hasContainer = false;
+          break;
+        }
+      }
+    }
+
+    let container;
+    if (hasContainer) {
+      container = body.firstElementChild;
+    } else {
+      container = doc.createElement('i-amp-container');
+      moveChildren(body, container);
+      body.appendChild(container);
+      // TODO Update any `body > X` CSS selectors to `body > i-amp-container > X`.
```

Comment:
This might be a hard TODO to satisfy since people might do the same via classes. E.g. `.mybody > div`. We can copy the classes from body to container, but that doesn't always work very well.